### PR TITLE
Add a plugin to provide autoimport functionality

### LINKED ISSUE #34: Auto-import?
jedi doesn't support auto-import: https://github.com/pappasam/jedi-language-server/issues/126#issuecomment-835878364
can python-lsp support it through other means?

### PR DESCRIPTION
**Pull request by bagel897**
# Autoimport plugin for pylsp
Depends on accompanying pull request in rope
@bagel897  
Closes #34
## Prereqs
Rope > 1.0.0 (AKA git builds rn)  
## Demo Images
image
can detect external modules
image
grabs names from the external modules
image

## Features
- [x] Score-based sorting algorithm with thresholding. Picks the 25 best options.
Will take suggestions on how to improve the basic algorithm
- [x] Relatively high preformance using sqllite3 and ast.
Besides the cache generation, it preforms fine on my machine
- [x] Avoids suggestions on single line comments, from/normal imports, function/classes, dots
Most of the test cases are to cover this and it should be pretty functional.
- [x] Most completion related tests should be in rope
the main ones are in parsing existing names and whether to suggest imports
- [x] inserts all imports at the bottom of the set of imports. Recommendation is to use the isort extension.
- [x] Watches for changes (but not deleted files) and regenerates cache for saved files.
### Won't fix
1. Correct detection of type of statements from  ``__init__.py`` modules. Would be difficult to preform type inference on.
image
2. Doesn't handle deleted files but this is a relatively rare.
3. Without a progress indicator, the user won't know when its ready (201) I have a WIP implementation based on #236 
4. Doesn't avoid suggestions on multiple line comments/strings, can fix if deemed important

### CODE REVIEW

File: pylsp/plugins/rope_autoimport.py
Comment (rchl) [inline pylsp/plugins/rope_autoimport.py]: Should be disabled by default probably? Both here and in the schema and documentation.

Comment (bagel897) [reply pylsp/plugins/rope_autoimport.py]: Done

File: CONFIGURATION.md
Comment (rchl) [inline CONFIGURATION.md]: if those are supposed to be in sorted order then should be flipped

**Comment by bagel897** [comment]
so this is almost ready:
1. It needs a fix for windows in rope https://github.com/python-rope/rope/pull/477 and associated release
2. Its still a little rough around where to provide suggestions and slow when it does. Ideally, some of this can be used in jedi and rope completions.
3. It generates a large cache which takes a while. Normally, this will be on-disk and I will work on improving it. But in CI, this takes a while and must be done seperately for each test and kept in memory. Either I change the scope of the associated fixtures to a module scope, or we have 2m long testing runs. In the future, I'll make it use configured dependencies automatically.

**Comment by bagel897** [comment]
Some notes about the future of this. I'm working on splitting it out to its own repository where 1-3 things will happen.
1. Rope-autoimport will use autoimport-core under the hood
2. Pylsp will use autoimport-core directly
3. Jedi will use autoimport-core, negating the requirements for any changes in pylsp.

But for now, this is perfectly functional code

**Comment by ccordoba12** [comment]
Hey @bageljrkhanofemus, thanks for your work on this! It looks really promising.

About what you said:

> I'm working on splitting it out to its own repository

We can invite you to our organization so you can create a repo for your plugin here and continue your work there. I'd say that's the best course of action.

**Comment by bagel897** [comment]
The plugin will stay here, but I'm trying to split out the underlying [library](https://github.com/bageljrkhanofemus/autoimport-core ). While it was developed *for* pylsp, I don't see why I'd need to put it on the pylsp organization.

**Comment by ccordoba12** [comment]
> While it was developed for pylsp, I don't see why I'd need to put it on the pylsp organization.

Some plugin authors prefer to maintain them in separate repos and I understood you were thinking on doing that.

**Comment by bagel897** [comment]
I can't because I must store the AutoImport object in the workspace (to persist in-memory databases), and normal plugins cannot do that. Ideally, that becomes a hook so they can. I also think having to install separate plugins is bad UX

**Comment by ccordoba12** [comment]
> I also think having to install separate plugins is bad UX

I understand, but the problem with all plugins being part of the core is that after they are merged, it's another thing that maintainers need to take care of, which can become overwhelming quite quickly.

**Comment by bagel897** [comment]
Fair, but unless there's a way to persist objects into workspace from plugins autoimport still needs to be here.
The biggest issue is that I wrote parso parsers in the plugin, but I don't know a better solution. I can bundle them into autoimport-core but we'd still be doing the AST parsing twice. Ideally, Jedi could handle that but that has its own issues.

**Comment by ccordoba12** [comment]
> Fair, but unless there's a way to persist objects into workspace from plugins autoimport still needs to be here.

Ok, no problem.

**Comment by wudstrand** [comment]
What's the status on this?

**Comment by bagel897** [comment]
> What's the status on this?


This PR needs to be reviewed. Possible concerns:
1. The tests timeout because they all index separately in memory - its not a cheap process to do repeatedly, but is a testing only issue. I have some ideas to fix that - namely dependency detection or making the autoimport object session scoped.
2. There's alot of parso code that I'd imagine the pylsp team understandably doesn't want to maintain here but I cannot put into rope
3. Progress support is in a separate PR. If that gets merged first, then I need to update this one or make a second PR.

I haven't heard that any are a blocker from the pylsp team so I can't really answer more.

**Comment by ccordoba12** [comment]
> The tests timeout because they all index separately in memory

You can skip them in Mac if they fail too much on that platform.

> There's alot of parso code that I'd imagine the pylsp team understandably doesn't want to maintain here but I cannot put into rope

That's fine as long as you commit to maintain it in the future.

> Progress support is in a separate PR

Does that PR need be merged too to finish your work to add auto-import support?

**Comment by bagel897** [comment]
- [x] Mac testing times - I made autoimport a session scoped fixture to reduce test timings. This should fix it for macs
- [x] Parsers - I think that the code I wrote should be similar to jedi's own code - I'm investigating if the Jedi API can replace some of my code. I've replaced the easier parser (get_names). The harder one (should_insert) is more specialized but jedi theoretically implements it.
- [x] Progress support - I have a second branch which I'll rebase when both PRs are merged, unless the progress PR gets merged first.

**Comment by bagel897** [comment]
> > There's alot of parso code that I'd imagine the pylsp team understandably doesn't want to maintain here but I cannot put into rope
> 
> That's fine as long as you commit to maintain it in the future.
 
As a student, I cannot reliably do so. I'll try to fix bugs but I make no guarantees on response time since academics have more precedence. You may find the same api in Jedi, but I couldn't find it. Rope cannot handle this since its ast based and it needs error-recovery.

**Comment by ccordoba12** [comment]
> As a student, I cannot reliably do so. I'll try to fix bugs but I make no guarantees on response time since academics have more precedence

Ok, I understand and no problem with that. The most important thing for us is that you're willing to address bugs related to this new functionality. The time you need to respond to them is up to you, of course.

**Comment by bagel897** [comment]
Yes, but its probably better to mark it off by default, and explain that it is experimental unless it is replaced in jedi/rope

**Comment by ccordoba12** [comment]
Agreed, that's fine for me too.

**Comment by mostley** [comment]
Can I help somehow to make this happen?

**Comment by bagel897** [comment]
> Can I help somehow to make this happen?

If Jedi manifested a should_insert API, that'd help - but isn't necessary for the merge. I'm not sure what else is needed.

**Comment by ccordoba12** [comment]
I didn't know this was ready. I'll try to review it this week so we can include it in our next version.

File: docs/autoimport.md
Comment (ccordoba12) [inline docs/autoimport.md]:
```suggestion
# Autoimport for pylsp

Requirements:
1. Install ``python-lsp-server[rope]``
2. Set ``pylsp.plugins.rope_autoimport.enabled`` to ``true`` 

## Startup

Autoimport will generate an autoimport sqllite3 database in .ropefolder/autoimport.db on startup.  
This will take a few seconds but should be much quicker on future runs.

## Usage 

Autoimport will provide suggestions to import names from everything in `sys.path`. You can change this by changing where pylsp is running or by setting rope's `python_path` option. 
It will suggest modules, submodules, keywords, functions, and classes.  

Since autoimport inserts everything towards the end of the import group, its recommended you use the isort [plugin](https://github.com/paradoxxxzero/pyls-isort).

## Credits
 - Most of the code was written by me, @bagel897 
```

File: pylsp/config/schema.json
Comment (ccordoba12) [inline pylsp/config/schema.json]:
```suggestion
    },
    "pylsp.plugins.rope_autoimport.enabled": {
      "type": "boolean",
      "default": false,
      "description": "Enable or disable autoimport."
    },
    "pylsp.plugins.rope_autoimport.memory": {
```

Please don't forget to update CONFIGURATION.md after this.

File: pylsp/plugins/rope_autoimport.py
Comment (ccordoba12) [inline pylsp/plugins/rope_autoimport.py:3]:
```suggestion
# Copyright 2022- Python Language Server Contributors.

import logging
```

File: pylsp/plugins/rope_autoimport.py
Comment (ccordoba12) [inline pylsp/plugins/rope_autoimport.py]:
```suggestion
```

Unnecessary blank line.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: - Please revert this change because the line length limit of this project is 120 columns.
- Also, please refrain of making this kind of changes in the future because they don't add anything significant to your work and make PRs harder to review.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pylsp/workspace.py
Comment (ccordoba12) [inline pylsp/workspace.py]: Please revert.

File: pyproject.toml
Comment (ccordoba12) [inline pyproject.toml]:
```suggestion
    "rope>1.2.0",
```

To match the requirement below.

File: test/plugins/test_autoimport.py
Comment (ccordoba12) [inline test/plugins/test_autoimport.py:3]:
```suggestion
# Copyright 2022- Python Language Server Contributors.

from typing import Dict, List
```

File: test/plugins/test_autoimport.py
Comment (ccordoba12) [inline test/plugins/test_autoimport.py]:
```suggestion
class TestShouldInsert:
```

We prefer snake-cased names for classes.

**Review by ccordoba12 — COMMENTED** [review]
Thanks a lot @bagel897 for your work on this! I left a very lightweight review for you, mostly about style issues. Otherwise looks good to me.

File: pylsp/plugins/rope_autoimport.py
Comment (bagel897) [reply pylsp/plugins/rope_autoimport.py:3]: Fixed in 76f68de3a9b56c9f3ee014bee0da0bd141f8547a

File: test/plugins/test_autoimport.py
Comment (bagel897) [reply test/plugins/test_autoimport.py:3]: Fixed in 76f68de3a9b56c9f3ee014bee0da0bd141f8547a

File: docs/autoimport.md
Comment (bagel897) [reply docs/autoimport.md]: Fixed in f5a2992d57c96e66191a04691f7947b01e28af27 (just used prettier)

File: pylsp/workspace.py
Comment (bagel897) [reply pylsp/workspace.py]: f5a2992d57c96e66191a04691f7947b01e28af27 (sorry for running black)

File: pyproject.toml
Comment (bagel897) [reply pyproject.toml]: Fixed in a8814981c6447dc6b352bbbb05db18c3df82aeb0

File: pylsp/plugins/rope_autoimport.py
Comment (bagel897) [reply pylsp/plugins/rope_autoimport.py]: 7f40c74

**Comment by bagel897** [comment]
Pushed a bugfix and the style changes. Let me know if there's anything else (or if the CI fails, I'll check). You might want to squash the PR, since its a lot of poorly labeled commits, but up to you. Thanks for the feedback @ccordoba12!

File: pylsp/plugins/rope_autoimport.py
Comment (ccordoba12) [inline pylsp/plugins/rope_autoimport.py]:
```suggestion
        if any(node.type == "operator" and "." in node.value or
               node.type == "trailer" for node in expr.children):
```

I think this should solve the style issue flagged by our CIs.

**Review by ccordoba12 — COMMENTED** [review]
Thanks for addressing my previous review @bagel897! One last suggestion for you then this should be ready.

**Comment by ccordoba12** [comment]
> You might want to squash the PR, since its a lot of poorly labeled commits, but up to you

Don't worry about that, we squash merge all PRs in this repo.

**Review by ccordoba12 — APPROVED** [review]
Thanks @bagel897 for this addition! It's a great improvement!!

**Pull request merged**
